File: gateway/src/main/java/bank/internalgateway/gateway/observability/RequestTraceService.java
package bank.internalgateway.gateway.observability;

import bank.internalgateway.gateway.config.GatewayProperties;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.ArrayList;
import java.util.DoubleSummaryStatistics;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.Collectors;

@Service
public class RequestTraceService {

    private final int maxEntries;
    private final int topRoutesLimit;
    private final ConcurrentLinkedDeque<RequestTraceEntry> entries = new ConcurrentLinkedDeque<>();
    private final AtomicLong totalRecorded = new AtomicLong();

    public RequestTraceService(GatewayProperties properties) {
        GatewayProperties.Observability observability = properties.observability();
        this.maxEntries = observability != null ? observability.traceBufferSize() : 500;
        this.topRoutesLimit = observability != null ? observability.topRoutesLimit() : 10;
    }

    public void record(RequestTraceEntry entry) {
        entries.addFirst(entry);
        totalRecorded.incrementAndGet();
        while (entries.size() > maxEntries) {
            entries.pollLast();
        }
    }

    public RequestTraceEntry start(String plane, String method, String inboundPath, String targetService, String targetUrl) {
        return new RequestTraceEntry(
                UUID.randomUUID().toString(),
                Instant.now(),
                plane,
                method,
                inboundPath,
                targetService,
                targetUrl,
                null,
                "IN_PROGRESS",
                0L,
                null
        );
    }

    public void complete(RequestTraceEntry entry, String status, long durationMs, String detail) {
        record(new RequestTraceEntry(
                entry.id(),
                entry.timestamp(),
                entry.plane(),
                entry.method(),
                entry.inboundPath(),
                entry.targetService(),
                entry.targetUrl(),
                entry.correlationId(),
                status,
                durationMs,
                detail
        ));
    }

    public void recordSimple(
            String plane,
            String method,
            String inboundPath,
            String targetService,
            String targetUrl,
            String correlationId,
            String status,
            long durationMs,
            String detail) {
        record(new RequestTraceEntry(
                UUID.randomUUID().toString(),
                Instant.now(),
                plane,
                method,
                inboundPath,
                targetService,
                targetUrl,
                correlationId,
                status,
                durationMs,
                detail
        ));
    }

    public List<RequestTraceEntry> recent(int limit) {
        int safeLimit = Math.max(1, Math.min(limit, maxEntries));
        return entries.stream().limit(safeLimit).toList();
    }

    public RequestStats stats() {
        List<RequestTraceEntry> snapshot = new ArrayList<>(entries);
        long success = snapshot.stream().filter(e -> e.status().startsWith("2") || "SUCCESS".equals(e.status())).count();
        long error = snapshot.stream().filter(e -> e.status().startsWith("4") || e.status().startsWith("5") || "ERROR".equals(e.status())).count();
        long dedup = snapshot.stream().filter(e -> "DEDUP".equals(e.status())).count();
        long rateLimited = snapshot.stream().filter(e -> "RATE_LIMIT".equals(e.status())).count();

        Map<String, Long> byPlane = snapshot.stream()
                .collect(Collectors.groupingBy(RequestTraceEntry::plane, Collectors.counting()));

        Map<String, Long> byRoute = snapshot.stream()
                .collect(Collectors.groupingBy(
                        e -> e.method() + " " + e.inboundPath(),
                        Collectors.counting()));

        List<RouteStat> topRoutes = byRoute.entrySet().stream()
                .sorted(Map.Entry.<String, Long>comparingByValue().reversed())
                .limit(topRoutesLimit)
                .map(e -> new RouteStat(e.getKey(), e.getValue()))
                .toList();

        DoubleSummaryStatistics latency = snapshot.stream()
                .filter(e -> e.durationMs() > 0)
                .mapToDouble(RequestTraceEntry::durationMs)
                .summaryStatistics();

        return new RequestStats(
                totalRecorded.get(),
                snapshot.size(),
                success,
                error,
                dedup,
                rateLimited,
                byPlane,
                topRoutes,
                latency.getCount() == 0 ? 0 : latency.getAverage(),
                latency.getCount() == 0 ? 0 : latency.getMax()
        );
    }

    public record RequestTraceEntry(
            String id,
            Instant timestamp,
            String plane,
            String method,
            String inboundPath,
            String targetService,
            String targetUrl,
            String correlationId,
            String status,
            long durationMs,
            String detail
    ) {
    }

    public record RouteStat(String route, long count) {
    }

    public record RequestStats(
            long totalRecorded,
            int bufferedEntries,
            long successCount,
            long errorCount,
            long dedupCount,
            long rateLimitedCount,
            Map<String, Long> byPlane,
            List<RouteStat> topRoutes,
            double avgDurationMs,
            double maxDurationMs
    ) {
    }
}

File: dotnet-gateway/src/InternalGateway.DotNet/Observability/RequestTraceService.cs
using System.Collections.Concurrent;
using Microsoft.Extensions.Options;

namespace InternalGateway.DotNet.Observability;

public sealed class RequestTraceService
{
    private readonly int _maxEntries;
    private readonly int _topRoutesLimit;
    private readonly ConcurrentQueue<RequestTraceEntry> _entries = new();
    private long _totalRecorded;

    public RequestTraceService(IOptions<GatewayOptions> options)
    {
        _maxEntries = options.Value.Observability.TraceBufferSize;
        _topRoutesLimit = options.Value.Observability.TopRoutesLimit;
    }

    public void RecordSimple(
        string plane,
        string method,
        string inboundPath,
        string? targetService,
        string? targetUrl,
        string? correlationId,
        string status,
        long durationMs,
        string? detail)
    {
        Record(new RequestTraceEntry(
            Guid.NewGuid().ToString("D"),
            DateTimeOffset.UtcNow,
            plane,
            method,
            inboundPath,
            targetService,
            targetUrl,
            correlationId,
            status,
            durationMs,
            detail));
    }

    public void Record(RequestTraceEntry entry)
    {
        _entries.Enqueue(entry);
        Interlocked.Increment(ref _totalRecorded);
        while (_entries.Count > _maxEntries && _entries.TryDequeue(out _))
        {
        }
    }

    public IReadOnlyList<RequestTraceEntry> Recent(int limit)
    {
        var take = Math.Max(1, limit);
        return _entries.Reverse().Take(take).ToArray();
    }

    public RequestStats Stats()
    {
        var snapshot = _entries.ToArray();
        var byRoute = snapshot
            .GroupBy(e => e.Method + " " + e.InboundPath, StringComparer.Ordinal)
            .Select(g => new RouteStat(g.Key, g.LongCount()))
            .OrderByDescending(r => r.Count)
            .Take(_topRoutesLimit)
            .ToArray();

        double Avg(Func<RequestTraceEntry, bool> pred)
        {
            var filtered = snapshot.Where(pred).Select(e => (double)e.DurationMs).ToArray();
            return filtered.Length == 0 ? 0 : filtered.Average();
        }

        return new RequestStats(
            Interlocked.Read(ref _totalRecorded),
            snapshot.Length,
            Avg(_ => true),
            Avg(e => e.Plane == "ingress"),
            Avg(e => e.Plane == "capability"),
            Avg(e => e.Plane == "messaging"),
            byRoute);
    }
}

public sealed record RequestTraceEntry(
    string Id,
    DateTimeOffset Timestamp,
    string Plane,
    string Method,
    string InboundPath,
    string? TargetService,
    string? TargetUrl,
    string? CorrelationId,
    string Status,
    long DurationMs,
    string? Detail);

public sealed record RouteStat(string Route, long Count);

public sealed record RequestStats(
    long TotalRecorded,
    int BufferedEntries,
    double AvgDurationMs,
    double AvgIngressMs,
    double AvgCapabilityMs,
    double AvgMessagingMs,
    IReadOnlyList<RouteStat> TopRoutes);
